src/canvastools/ts/CanvasTools/Region/Component/MidpointComponent.ts

```diff
@@ -0,0 +1,141 @@
+    super(paper, paperRect, regionData, callbacks);
+    this.node = paper.g();
+    this.node.addClass("midpointLayer");
+    this.midpointElements = [];
+    this.midpointNode = this.paper.g();
+    this.node.add(this.midpointNode);
+    const regionMidpoints = this.regionData.getLineMidpoints();
+    this.buildMidpoints(regionMidpoints);
+  }
+
+  /**
+   * Helper function to create a new midpoint.
+   * @param paper - The `Snap.Paper` object to draw on.
+   * @param x - The `x`-coordinate of the midpoint.
+   * @param y - The `y`-coordinate of the midpoint.
+   * @param style - Additional css style class to be applied.
+   * @param r - The radius of the midpoint.
+   */
+  protected createMidpoint(
+    paper: Snap.Paper,
+    x: number,
+    y: number,
+    style?: string,
+    r: number = MidpointComponent.DEFAULT_RADIUS
+  ): Snap.Element {
+    const midpoint = paper.circle(x, y, r);
+    midpoint.addClass("midpointStyle");
+    if (style !== undefined && style !== "") {
+      midpoint.addClass(style);
+    }
+    return midpoint;
+  }
+
+  protected teardownMidpoints() {
+    this.midpointElements.forEach(midpointElement => {
+      midpointElement.remove();
+    });
+    this.midpointElements = [];
+  }
+
+  protected buildMidpoints(regionMidpoints: Point2D[]) {
+    this.teardownMidpoints();
+    regionMidpoints.forEach((point, index) => {
+      const midpoint = this.createMidpoint(this.paper, point.x, point.y);
+      this.midpointElements.push(midpoint);
+      this.midpointNode.add(midpoint);
+      this.subscribeMidpointToEvents(midpoint, index);
+    });
+  }
+
+  public redraw() {
+    const regionMidpoints = this.regionData.getLineMidpoints();
+    if (this.midpointElements.length !== regionMidpoints.length) {
+      // if # of midpoints has changed, rebuild them
+      window.requestAnimationFrame(() => {
+        this.buildMidpoints(regionMidpoints);
+      });
+    } else {
+      // update midpoints in place
+      window.requestAnimationFrame(() => {
+          regionMidpoints.forEach((p, index) => {
+              this.midpointElements[index].attr({
+                  cx: p.x,
+                  cy: p.y,
+              });
+          });
+      });
+    }
+  }
+
+  /**
+   * Add event listeners to a midpoint's DOM node
+   */
+  protected subscribeMidpointToEvents(midpoint: Snap.Element, index: number) {
+    const listeners: EventListeners = [
+      {
+        event: "pointerenter",
+        base: midpoint.node,
+        listener: (e: PointerEvent) => {
+          e.stopPropagation();
+          console.log(`Midpoint entered. Index = ${index}`);
+          this.activeMidpointIndex = index;
+        },
+        bypass: false
+      },
+      {
+        event: "click",
+        base: midpoint.node,
+        listener: (e: MouseEvent) => {
+          e.stopPropagation();
+          console.log("Midpoint component clicked", e);
+        },
+        bypass: false,
+      },
+    ];
+    this.subscribeToEvents(listeners);
```

Comment:
If we end up calling the function twice then we subscribe twice? Should/can we make idempotent?

Reply:
I think these subscribe calls are only called in constructors, so any new nodes have listeners attached to them. I'll double check.